File: tools/v41/enc_check.cpp
// enc_check — the engine's V4.1 chat encoding for a few fixed conversations, one per line (escaped),
// to diff against DeepSeek-V4.1's encoding.py (tools/v41/enc_check.py prints the reference side).
#include "engine/chat_encode.h"
#include <cstdio>
#include <string>
#include <vector>
using namespace aff;
static void emit(const std::vector<ChatMsg>& m, ThinkingMode t, int b) {
  EncodeOpts o; o.thinking_mode = t; o.v41 = true; o.effort_budget = b;
  const std::string s = encode_messages(m, o);
  for (char c : s) { if (c == '\n') std::fputs("\\n", stdout); else std::fputc(c, stdout); }
  std::fputc('\n', stdout);
}
int main() {
  ChatMsg u; u.role = "user"; u.content = "What is the capital of France?";
  ChatMsg sy; sy.role = "system"; sy.content = "You are terse.";
  ChatMsg a; a.role = "assistant"; a.content = "Paris."; a.reasoning_content = "Easy.";
  ChatMsg u2; u2.role = "user"; u2.content = "And Spain?";
  emit({u}, ThinkingMode::Chat, 100);
  emit({u}, ThinkingMode::Thinking, 100);
  emit({u}, ThinkingMode::Thinking, 75);
  emit({sy, u}, ThinkingMode::Chat, 100);
  emit({sy, u}, ThinkingMode::Thinking, 50);
  emit({u, a, u2}, ThinkingMode::Thinking, 100);
  emit({u, a, sy, u2}, ThinkingMode::Chat, 100);
}

File: tools/v41/enc_check.py
#!/usr/bin/env python3
"""The reference side of enc_check.cpp: the same conversations through DeepSeek-V4.1's encoding.py."""
import importlib.util, sys
p = sys.argv[1] if len(sys.argv) > 1 else "/path/to/hf/DeepSeek-V4.1-Flash/encoding/encoding.py"
s = importlib.util.spec_from_file_location("e41", p); e = importlib.util.module_from_spec(s); s.loader.exec_module(e)
u = {"role": "user", "content": "What is the capital of France?"}
sy = {"role": "system", "content": "You are terse."}
a = {"role": "assistant", "content": "Paris.", "reasoning_content": "Easy."}
u2 = {"role": "user", "content": "And Spain?"}
for msgs, mode, eff in [([u], "chat", 100), ([u], "thinking", 100), ([u], "thinking", 75), ([sy, u], "chat", 100),
                        ([sy, u], "thinking", 50), ([u, a, u2], "thinking", 100), ([u, a, sy, u2], "chat", 100)]:
    print(e.encode_messages(msgs, thinking_mode=mode, reasoning_effort=eff).replace("\n", "\\n"))
